### PR TITLE
Using common subsctription code with TNT impl.

### PR DESCRIPTION
**Pull request by douradofrazer**

### CODE REVIEW

File: src/main/java/org/dcsa/tnt/service/impl/TNTEventSubscriptionTOServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+  public Mono<TNTEventSubscriptionTO> update(TNTEventSubscriptionTO eventSubscriptionTO) {
+    if (eventSubscriptionTO.getSecret() != null) {
+      return Mono.error(
+          new UpdateException(
+              "Please omit the \"secret\" attribute.  If you want to change the"
+                  + " secret, please use the dedicated secret endpoint"
+                  + " (\"PUT .../event-subscriptions/"
+                  + eventSubscriptionTO.getSubscriptionID()
+                  + "/secret\")."));
+    }
```
Comment (nt-as) [inline src/main/java/org/dcsa/tnt/service/impl/TNTEventSubscriptionTOServiceImpl.java:85]: Could this be done in the subclass or a preUpdateHook instead, so we do not have to repeat it in every implementation?

File: src/main/java/org/dcsa/tnt/service/impl/TNTEventSubscriptionTOServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+        .flatMap(
+            td ->
+                eventSubscriptionRepository.insertTransportDocumentEventTypeForSubscription(
+                    eventSubscriptionTO.getSubscriptionID(), td))
+        .then();
+  }
+
+  private Mono<Void> createShipmentEventType(TNTEventSubscriptionTO eventSubscriptionTO) {
+
+    if (null == eventSubscriptionTO.getShipmentEventTypeCode()) return Mono.empty().then();
```
Comment (nt-as) [inline src/main/java/org/dcsa/tnt/service/impl/TNTEventSubscriptionTOServiceImpl.java:140]: I think we should map null to all of them

File: src/main/java/org/dcsa/tnt/service/impl/TNTEventSubscriptionTOServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+        .flatMap(
+            s ->
+                eventSubscriptionRepository.insertShipmentEventTypeForSubscription(
+                    eventSubscriptionTO.getSubscriptionID(), s))
+        .then();
+  }
+
+  private Mono<Void> createTransportEventType(TNTEventSubscriptionTO eventSubscriptionTO) {
+
+    if (null == eventSubscriptionTO.getTransportEventTypeCode()) return Mono.empty().then();
```
Comment (nt-as) [inline src/main/java/org/dcsa/tnt/service/impl/TNTEventSubscriptionTOServiceImpl.java:155]: I think we should map null to all of them

File: src/main/java/org/dcsa/tnt/service/impl/TNTEventSubscriptionTOServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+        .flatMap(
+            t ->
+                eventSubscriptionRepository.insertTransportEventTypeForSubscription(
+                    eventSubscriptionTO.getSubscriptionID(), t))
+        .then();
+  }
+
+  private Mono<Void> createEquipmentEventType(TNTEventSubscriptionTO eventSubscriptionTO) {
+
+    if (null == eventSubscriptionTO.getEquipmentEventTypeCode()) return Mono.empty().then();
```
Comment (nt-as) [inline src/main/java/org/dcsa/tnt/service/impl/TNTEventSubscriptionTOServiceImpl.java:169]: I think we should map null to all of them

File: src/main/java/org/dcsa/tnt/service/impl/TNTEventSubscriptionTOServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+        .concatMap(
+            eventType ->
+                eventSubscriptionRepository.insertEventTypeForSubscription(
+                    eventSubscriptionTO.getSubscriptionID(), eventType))
+        .then(Mono.just(eventSubscriptionTO));
+  }
+
+  private Mono<Void> createTransportDocumentEventTypes(TNTEventSubscriptionTO eventSubscriptionTO) {
+
+    if (null == eventSubscriptionTO.getTransportDocumentTypeCode()) return Mono.empty().then();
```
Comment (nt-as) [inline src/main/java/org/dcsa/tnt/service/impl/TNTEventSubscriptionTOServiceImpl.java:126]: I think we should map null to all of them

File: src/main/java/org/dcsa/tnt/controller/TNTEventSubscriptionTOController.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    return super.create(tntEventSubscriptionTO);
+  }
+
+  @Override
+  public Mono<TNTEventSubscriptionTO> update(
+      @PathVariable UUID id,
+      @Validated(value = ValidationGroups.Update.class) @RequestBody
+          TNTEventSubscriptionTO tntEventSubscriptionTO) {
+    return super.update(id, tntEventSubscriptionTO);
+  }
```
Comment (HenrikHL) [inline src/main/java/org/dcsa/tnt/controller/TNTEventSubscriptionTOController.java:50]: Why is this needed? Is it because of `ValidationGroups.Create.class`?

File: src/main/java/org/dcsa/tnt/service/impl/TNTEventSubscriptionTOServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+      eventTypes = eventSubscriptionTO.getEventType();
+    }
+
+    return Flux.fromIterable(eventTypes)
+        .concatMap(
+            eventType ->
+                eventSubscriptionRepository.insertEventTypeForSubscription(
+                    eventSubscriptionTO.getSubscriptionID(), eventType))
+        .then(Mono.just(eventSubscriptionTO));
+  }
```
Comment (HenrikHL) [inline src/main/java/org/dcsa/tnt/service/impl/TNTEventSubscriptionTOServiceImpl.java:124]: To me this should be located in EventCore with an abstract method called getAllAllowedEventTypes() which needs to be implemented in this class

File: src/main/java/org/dcsa/tnt/service/impl/TNTEventSubscriptionTOServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+
+    List<TransportDocumentTypeCode> transportDocumentTypeCodes =
+        eventSubscriptionTO.getTransportDocumentTypeCode();
+    return Flux.fromIterable(transportDocumentTypeCodes)
+        .flatMap(
+            td ->
+                eventSubscriptionRepository.insertTransportDocumentEventTypeForSubscription(
+                    eventSubscriptionTO.getSubscriptionID(), td))
+        .then();
+  }
```
Comment (HenrikHL) [inline src/main/java/org/dcsa/tnt/service/impl/TNTEventSubscriptionTOServiceImpl.java:140]: This method will also be needed in OVS - why not put it in EventCore? and refer to an abstract function getAllTransportDocumentTypes()

File: src/main/java/org/dcsa/tnt/service/impl/TNTEventSubscriptionTOServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+
+    List<TransportEventTypeCode> transportEventTypeCode =
+        eventSubscriptionTO.getTransportEventTypeCode();
+    return Flux.fromIterable(transportEventTypeCode)
+        .flatMap(
+            t ->
+                eventSubscriptionRepository.insertTransportEventTypeForSubscription(
+                    eventSubscriptionTO.getSubscriptionID(), t))
+        .then();
+  }
```
Comment (HenrikHL) [inline src/main/java/org/dcsa/tnt/service/impl/TNTEventSubscriptionTOServiceImpl.java:173]: This is also needed in OVS

File: src/main/java/org/dcsa/tnt/service/impl/TNTEventSubscriptionTOServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+    List<ShipmentEventTypeCode> shipmentEventTypeCode =
+        eventSubscriptionTO.getShipmentEventTypeCode();
+
+    return Flux.fromIterable(shipmentEventTypeCode)
+        .flatMap(
+            s ->
+                eventSubscriptionRepository.insertShipmentEventTypeForSubscription(
+                    eventSubscriptionTO.getSubscriptionID(), s))
+        .then();
+  }
```
Comment (HenrikHL) [inline src/main/java/org/dcsa/tnt/service/impl/TNTEventSubscriptionTOServiceImpl.java:157]: Even though this is not needed in OVS - it is needed in Documentation_Hub - I think it belongs to EventCore

File: src/main/java/org/dcsa/tnt/service/impl/TNTEventSubscriptionTOServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+    List<EquipmentEventTypeCode> equipmentEventTypeCode =
+        eventSubscriptionTO.getEquipmentEventTypeCode();
+
+    return Flux.fromIterable(equipmentEventTypeCode)
+        .flatMap(
+            e ->
+                eventSubscriptionRepository.insertEquipmentEventTypeForSubscription(
+                    eventSubscriptionTO.getSubscriptionID(), e))
+        .then();
+  }
```
Comment (HenrikHL) [inline src/main/java/org/dcsa/tnt/service/impl/TNTEventSubscriptionTOServiceImpl.java:190]: Put in EventCore - same argument as above - we might need it in the future (DCSA_EVENT_HUB)

File: src/main/java/org/dcsa/tnt/controller/TNTEventSubscriptionTOController.java (hunk 6 above)
Comment (douradofrazer) [reply src/main/java/org/dcsa/tnt/controller/TNTEventSubscriptionTOController.java:50]: With this we can specify which validations need to be check based on the group mentioned. 
since we can have different validations for POST and PUT (Which is the case here) it will allow us to specify the validations based on which method is being called.
For example, in POST `secret` is required but in PUT  `secret` needs to be null.

**Review by HenrikHL — APPROVED** [review]
Lets go with this
